sound/soc/sof/intel/cnl.c

```diff
@@ -29,6 +30,65 @@ static const struct snd_sof_debugfs_map cnl_dsp_debugfs[] = {
 static void cnl_ipc_host_done(struct snd_sof_dev *sdev);
 static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev);
 
+irqreturn_t cnl_ipc4_irq_thread(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+	bool ipc_irq = false;
+	u32 hipcida;
+	u32 hipctdr;
+	u32 hipctdd;
+
+	hipcida = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDA);
+	hipctdr = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCTDR);
+	hipctdd = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCTDD);
+
+	if (hipcida & CNL_DSP_REG_HIPCIDA_DONE)
+		/* mask Done interrupt */
+		snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCCTL,
+					CNL_DSP_REG_HIPCCTL_DONE, 0);
+
+	/* new message from DSP */
+	if (hipctdr & CNL_DSP_REG_HIPCTDR_BUSY) {
+		u32 msg = hipctdr & CNL_DSP_REG_HIPCTDR_MSG_MASK;
+		u32 msg_ext = hipctdd & CNL_DSP_REG_HIPCTDD_MSG_MASK;
+
+		if (hipctdr & SOF_IPC4_MSG_DIR_MASK) {
+			struct sof_ipc4_msg *data = sdev->ipc->msg.reply_data;
+
+			data->primary = msg;
+			data->extension = msg_ext;
+
+			spin_lock_irq(&sdev->ipc_lock);
+
+			snd_sof_ipc_get_reply(sdev);
+			snd_sof_ipc_reply(sdev, data->primary);
+
+			cnl_ipc_dsp_done(sdev);
+
+			spin_unlock_irq(&sdev->ipc_lock);
+		} else {
+			struct sof_ipc4_msg data = {{ 0 }};
+
+			data.primary = msg;
+			data.extension = msg_ext;
+
+			sdev->ipc->msg.rx_data = &data;
+			snd_sof_ipc_msgs_rx(sdev);
+			sdev->ipc->msg.rx_data = NULL;
+		}
+
+		cnl_ipc_host_done(sdev);
```

Comment:
Just curious, do we need spin_lock_irq(&sdev->ipc_lock); for cnl_ipc_host_done(sdev);?

Reply:
We were not doing that in IPC3, do you think we should in IPC4?

Reply:
> We were not doing that in IPC3, do you think we should in IPC4?

No, my question was for both IPC3 and IPC4. I don't know if it is possible that something can interleave when we set the busy and done bits.

Reply:
@bardliao, What I know that we can receive notification between sending a message and getting back the reply. I believe we have the higher level serialization.
we must not block the rx, which is async in many ways.

So far it worked fine, but we can take a deeper look and address both IPCs if there is a need?

Reply:
> @bardliao, What I know that we can receive notification between sending a message and getting back the reply. I believe we have the higher level serialization. we must not block the rx, which is async in many ways.
> 
> So far it worked fine, but we can take a deeper look and address both IPCs if there is a need?

Sure

Reply:
@bardliao, I think the IPC4 logic is not correct. I'll update the PR when I figured it out, but in short:
`(hipcida & CNL_DSP_REG_HIPCIDA_DONE)` indicates that the DSP acked a tx, nothing else.
`(hipctdr & CNL_DSP_REG_HIPCTDR_BUSY)` indicates that the DSP send a reply _or_ a new notification.

Reply:
good catch @ujfalusi